src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt

```diff
@@ -108,6 +146,173 @@ class MangaReaderViewModel : ViewModel() {
         return null
     }
 
+    // Function which will return True if format changes to 'reverse' format
+    private fun isChangedToManga() : Boolean {
+        return currentReaderFormat == READER_FORMAT_MANGA && wasReaderFormat != READER_FORMAT_MANGA
+    }
+
+    // Function which will return True if format changes from 'reverse' format to some else
+    private fun isChangedToBook() : Boolean {
+        return currentReaderFormat != READER_FORMAT_MANGA && wasReaderFormat == READER_FORMAT_MANGA
+    }
+
+    // Function which will redraw viewPager2 with pages with correct chosen format
+    fun redrawMangaReader() {
+        // change orientation
+        if (currentReaderFormat == READER_FORMAT_MANHWA) {
+            mangaReaderVP2.orientation = ViewPager2.ORIENTATION_VERTICAL
+        }
+        else
+            mangaReaderVP2.orientation = ViewPager2.ORIENTATION_HORIZONTAL
+
+        // change viewpager position
+        if (isChangedToManga() || isChangedToBook()) {
+            var pageIndex = manga.chapters[manga.lastViewedChapter].lastViewedPage
+            var delta = if (isChangedToBook()) {
+                pageIndex = (pagesCount - 1) - pageIndex
+                when {
+                    isSingleChapterManga() -> 0
+                    isOnFirstChapter() -> 1
+                    isOnLastChapter() -> 0
+                    else -> 1
+                }
+            }
+            else
+                when {
+                    isSingleChapterManga() -> 0
+                    isOnFirstChapter() -> 0
+                    isOnLastChapter() -> 1
+                    else -> 1
+                }
+
+            mangaReaderVP2.setCurrentItem(
+                (mangaReaderVP2.adapter!!.itemCount - 1) - pageIndex - delta,
+                false)
+        }
+
+        val pagerAdapter = mangaReaderVP2.adapter
+        pagerAdapter?.notifyDataSetChanged()
+        wasReaderFormat = currentReaderFormat
+    }
+
+    // Function which will load previous chapter after scroll
+    fun doToPrevChapter(viewPager : ViewPager2, pagerAdapter : MangaReaderVPAdapter) {
+        // update chapter
+        manga.lastViewedChapter--
+
+        // update pages count (and load chapter)
+        try {
+            pagesCount = manga
+                .chapters[manga.lastViewedChapter].getPagesNum()
+        }
+        catch (ex:MangaJetException) {
+            Toast.makeText(MangaJetApp.context, ex.message, Toast.LENGTH_SHORT).show()
+            viewPager.setCurrentItem(1, false)
+            manga.lastViewedChapter++
+            return
+        }
+
+        // start loading all pages
+        val job = GlobalScope.launch(Dispatchers.IO) {
+            uploadPages()
```

Comment:
Looks like you create a thread to... cancel all bitmap-threads?
And wait for it?
I think this it bad idea, because job.cancel (which you use in uploadPages) are
non-blocking. So you can do it just in main scope at nothing will be wrong